src/app/digital/public/api/impl/DigitalComponent.ts

```diff
@@ -16,4 +18,34 @@ export class DigitalComponentImpl extends ComponentImpl<
     public override get info(): DigitalComponentInfo {
         return new DigitalComponentInfoImpl(this.circuit, this.kind);
     }
+
+    public override firstAvailable(portGroup: Port["group"]): DigitalPort | undefined {
+        const ports = this.internal.getPortsForComponent(this.id);
+
+        const inputTypes = this.info.inputPortGroups;
+        const outputTypes = this.info.outputPortGroups;
+
+        const isInputType = inputTypes.includes(portGroup);
+        const isOutputType = outputTypes.includes(portGroup)
+
+        if (!isInputType && !isOutputType) 
+            return undefined;
+
+        function firstAvailableHelper(comp: DigitalComponentImpl, id: string): Boolean {
+            let portObject = comp.internal.getPortByID(id);
+            let portWire = comp.internal.getWiresForPort(id); // no wires = available
+    
+            if (comp.info.inputPortGroups && portObject !== undefined && portObject.group === portGroup)
```

Comment:
instead of checking `if (portObject !== undefined)` both times, above this do
```ts
if (!portObject)
    return false;
```
But I think you can safely assume it will always be defined, so up at `portObject` so do
`const portObject = comp.internal.getPortByID(id)!;` (note the `!` which asserts that it's defined)